src/pgen/cluster/agn_triggering.cpp

```diff
@@ -0,0 +1,626 @@
+
+    switch (agn_triggering.triggering_mode_) {
+    case AGNTriggeringMode::COLD_GAS: {
+
+      triggering_file << hydro_pkg->Param<Real>("agn_triggering_cold_mass");
+      break;
+    }
+    case AGNTriggeringMode::BOOSTED_BONDI:
+    case AGNTriggeringMode::BOOTH_SCHAYE: {
+      const auto &total_mass = hydro_pkg->Param<Real>("agn_triggering_total_mass");
+      const auto &avg_density =
+          hydro_pkg->Param<Real>("agn_triggering_mass_weighted_density") / total_mass;
+      const auto &avg_velocity =
+          hydro_pkg->Param<Real>("agn_triggering_mass_weighted_velocity") / total_mass;
+      const auto &avg_cs =
+          hydro_pkg->Param<Real>("agn_triggering_mass_weighted_cs") / total_mass;
+      triggering_file << total_mass << " " << avg_density << " " << avg_velocity << " "
+                      << avg_cs;
+      break;
+    }
+    case AGNTriggeringMode::NONE: {
+      break;
+    }
+    }
+
+    triggering_file << std::endl;
+    triggering_file.close();
+  }
+
+  // Remove accreted gas if using a Bondi-like mode
+  if (agn_triggering.remove_accreted_mass_) {
+    switch (agn_triggering.triggering_mode_) {
+    case AGNTriggeringMode::BOOSTED_BONDI:
+    case AGNTriggeringMode::BOOTH_SCHAYE: {
+      auto fluid = hydro_pkg->Param<Fluid>("fluid");
+      if (fluid == Fluid::euler) {
+        agn_triggering.RemoveBondiAccretedGas(md, tm.dt,
+                                              hydro_pkg->Param<AdiabaticHydroEOS>("eos"));
+      } else if (fluid == Fluid::glmmhd) {
+        agn_triggering.RemoveBondiAccretedGas(
+            md, tm.dt, hydro_pkg->Param<AdiabaticGLMMHDEOS>("eos"));
+      } else {
+        PARTHENON_FAIL("AGNTriggeringFinalizeTriggering: Unknown EOS");
+      }
+      break;
+    }
+    case AGNTriggeringMode::COLD_GAS: // Already removed during reduction
+    case AGNTriggeringMode::NONE: {
+      break;
+    }
+    }
+  }
+
+  return TaskStatus::complete;
+}
+
+// Limit timestep to a factor of the cold gas accretion time for Cold Gas
+// triggered cooling, or a factor of the time to accrete the total mass for
+// Bondi-like accretion
+parthenon::Real
+AGNTriggering::EstimateTimeStep(parthenon::MeshData<parthenon::Real> *md) const {
+  auto hydro_pkg = md->GetBlockData(0)->GetBlockPointer()->packages.Get("Hydro");
+
+  switch (triggering_mode_) {
+  case AGNTriggeringMode::COLD_GAS: {
+    return accretion_cfl_ * cold_t_acc_;
+  }
+  case AGNTriggeringMode::BOOSTED_BONDI:
+  case AGNTriggeringMode::BOOTH_SCHAYE: {
+    // Test the Bondi-like triggering methods
+    const Real total_mass = hydro_pkg->Param<Real>("agn_triggering_total_mass");
+    if (total_mass == 0) {
+      // TODO(forrestglines)During the first timestep, the total mass and
+      // accretion rate has not yet been reduced. However, since accreted mass is
+      // removed during that reduction, the timestep is needed to execute that
+      // reduction. As a compromise, we ignore the timestep constraint during the
```

Comment:
Isn't this even more general than just the first timestep?
I.e., if `total_mass==0` nothing should happen independent where in the simulation we are, doesn't it?

Reply:
If `total_mass==0` after the first reduction, then the total mass in the accretion zone is 0 and something must be broken. Only between its initialization and the first reduction should `total_mass==0`. Nonzero `total_mass` should carry over with restarts.